### ISSUE TITLE
Simplify input type definition

### ISSUE THREAD

**Issue by tgriesser** [graphql-nexus/schema#638]
One of the things I'd like to consider before 1.0 is whether there's a simpler way for us to express `inputObjectType`.

The only reason I originally designed the current API to match output, was to be able to capture `TypeName` / `FieldName` as generics to know the type for a potential `defaultArg`, which is actually a really rare use case, dead simple to remedy, and should be validated by `graphql-js` at runtime of the schema definition anyway. 

The current API is a lot of ceremony for not nearly the amount of gain you get with typing on the output end with arg definition, the `resolve` fn params & return value, any typing on plugins like `authorize`, etc. Input fields don't need any of that.

Further, it's very common to re-use field across input types... being able to define a decent base "partial" of an input type, and be able to spread it across multiple input types is very useful, the way you can with `args`.

One idea, is a new leaner API `inputType` to live alongside the existing `inputObjectType`, with a new `inputField` if you want to decorate w/ `description` / `deprecation` / `default`:

```ts
inputType('InputTypeName', {
  someRequiredType: nonNull('SomeInputType'),
  someString: 'String',
  someInt: 'Int',
  requiredInt: nonNull('Int'),
  listWithDescription: inputField(list('Int'), { description:  "This has a description" }),
  someType: OtherInputType
})
```

compare that to:

```ts
inputObjectType({
  name: 'InputTypeName',
  definition(t) {
    t.field('someRequiredField', {
      type: 'SomeInputType'
    })
    t.string('someString')
    t.string('someInt')
    t.nonNull.string('requiredInt')
    t.list.int('listWithDescription', {
      description: "This has a description"
    })
    t.field(OtherInputType)
  }
})
```

I think the ability to define & merge fields across multiple input types is the really powerful point here. What would be even better though, is if we could make `args` & `inputType` members have a unified API, as they do in `graphql-js`.

Not 100% convinced on this / not really suggesting we remove the existing `inputObjectType`, but now that `nullable / nonNull / list` api has landed - this was a big thing I was imagining that API could be really useful for.

@jasonkuhrt / @Weakky WDYT?

**Comment by jasonkuhrt** [comment]
> I think the ability to define & merge fields across multiple input types is the really powerful point here.

Agreed

I like.

Right now it is possible for plugins to enhance the type builder api `t` with new methods and properties. I haven't personally seen interesting use-cases yet for this but it is there.

It seems that we lose the ability for plugins to extend anything to do with `inputType`. That said, it is arguable that its not needed in the first place since now you can just spread, use pure functions, ... etc.

**Comment by Nayni** [comment]
What I don't really like is that `inputType` now looks completely different from a regular `objectType`. This means I now have to learn 2 ways of doing things instead of 1. I actually like that inputs look similar to an objectType in terms of definition.

> Right now it is possible for plugins to enhance the type builder api `t` with new methods and properties. I haven't personally seen interesting use-cases yet for this but it is there.

While I can't show actual code I have to say here that this has been a super powerful feature for my team. Being able to extend nexus without our builders on the `t` builder opens up some very powerful features.

One feature we use this on is we've built something amongst the lines of a `t.localised` which actually in the schema makes a regular `string` as output but builds a field with automatic `args` and a default implementation that knows how to localise the string in the current user's language. Being able to build that with a simple plugin is awesome.

**Comment by tgriesser** [comment]
> What I don't really like is that inputType now looks completely different from a regular objectType. This means I now have to learn 2 ways of doing things instead of 1. I actually like that inputs look similar to an objectType in terms of definition.

Not suggesting that we would remove `inputObjectType`, just that `inputType` would also. be available for situations where you might want to easily merge fields across multiple types via object spread, or for simpler syntax when defining "inline" fields

> While I can't show actual code I have to say here that this has been a super powerful feature for my team. Being able to extend nexus with custom builders on the t builder opens up some very powerful features.

Cool to hear!

One other idea before 1.0 is seeing if we can make a core plugin for validating input args ahead of execution. I think I might have remembered someone saying they'd been working on one.

**Comment by Nayni** [comment]
> Not suggesting that we would remove `inputObjectType`, just that `inputType` would also. be available for situations where you might want to easily merge fields across multiple types via object spread, or for simpler syntax when defining "inline" fields

Yeah I totally see the simplicity, my main concern is just confusion for new people. I personally prefer a consistent api that might be a little more explicit over having multiple different ways that I all have to learn, but maybe that's a bit of a personal preference.

**Comment by Sytten** [comment]
I also agree with @Nayni on that. One syntax is better than two, I feel this simplified syntax could easily be a plugin. I don't think the core schema should try to do too much. Even the mix and match of object and function in the `args` part of `objectType` is already pretty annoying IMO (`intArg` vs `t.int()`). I generally prefer the object way vs the function way, but I can understand thats not a popular in the nexus world.

**Comment by tgriesser** [comment]
> I feel this simplified syntax could easily be a plugin.

Yeah that's actually a very good point, I was on the fence about this to begin with. Agreed this could probably be a userland thing / show up in the docs as an example of what you could do.

> intArg vs t.int()

Yeah, makes sense. The difference here is that `t.int` is a builder for the object's fields, whereas `intArg` isn't bound by use, the same `arg` def can be used in multiple positions. 

In theory, we could add `t.intArg` but it'd just be sugar for the import (another thing that could be a plugin 😄

**Comment by jasonkuhrt** [comment]
Just hit a use-case for this while doing some design sketching for the new Nexus Prisma Plugin
